src/raftstore/store/snap/builder.rs

```diff
@@ -0,0 +1,465 @@
+    core: Arc<SnapManagerCore>,
+    cf_index: usize,
+    success: bool,
+}
+
+impl SnapshotReceiver {
+    /// Check the snapshot is on disk and correct first.
+    /// If not, open tmp files to receive it from network.
+    pub(super) fn new(
+        dir: String,
+        key: SnapKey,
+        io_limiter: Option<Arc<IOLimiter>>,
+        snapshot_meta: SnapshotMeta,
+        core: Arc<SnapManagerCore>,
+    ) -> Result<Self> {
+        let inner = Inner::new(dir, false, key, io_limiter).map_err(|e| {
+            let mut registry = core.apply_registry.lock().unwrap();
+            registry.remove(&key);
+            e
+        })?;
+
+        Ok(SnapshotReceiver {
+            inner,
+            snapshot_meta,
+            core,
+            cf_index: 0,
+            success: false,
+        })
+    }
+
+    pub fn save(&mut self) -> Result<()> {
+        let cf_files = self.inner.save()?.take_cf_files().into_iter();
+        for (got_cf, want_cf) in cf_files.zip(self.snapshot_meta.get_cf_files()) {
+            let expected = want_cf.get_size();
+            let got = got_cf.get_size();
+            if expected != got {
+                return Err(snapshot_size_corrupt(expected, got));
+            }
+
+            let expected = want_cf.get_checksum();
+            let got = got_cf.get_checksum();
+            if expected != got {
+                return Err(snapshot_checksum_corrupt(expected, got));
+            }
+        }
+        self.success = true;
+        let size = get_size_from_snapshot_meta(&self.snapshot_meta);
+        self.core.snap_size.fetch_add(size, Ordering::SeqCst);
+
+        let mut registry = self.core.apply_registry.lock().unwrap();
+        let entry = registry.get_mut(&self.inner.key).unwrap();
+        entry.meta = Some(self.snapshot_meta.clone());
+        Ok(())
+    }
+}
+
+impl Drop for SnapshotReceiver {
+    fn drop(&mut self) {
+        let mut registry = self.core.apply_registry.lock().unwrap();
+        if !self.success {
+            info!("{} receive snapshot fail, deregister", self.inner.key);
+            registry.remove(&self.inner.key);
+        }
+    }
+}
+
+impl Write for SnapshotReceiver {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if buf.is_empty() {
+            return Ok(0);
+        }
+        let inner = &mut self.inner;
+        while self.cf_index < SNAPSHOT_CFS.len() {
+            let cf_meta = &self.snapshot_meta.get_cf_files()[self.cf_index];
+
+            if self.cf_index >= inner.tmp_cf_files.len() {
+                let cf = &SNAPSHOT_CFS[self.cf_index];
+                let (dir, for_send, key) = (&inner.dir, inner.for_send, inner.key);
+                let path = get_cf_tmp_file_path(dir, for_send, key, cf);
+                let file = OpenOptions::new().write(true).create_new(true).open(&path)?;
+                let cf_file = CfFile::new(cf, path, Either::Left(file));
+                inner.tmp_cf_files.push(cf_file);
+            }
```

Comment:
Please extract above logic and below `self.cf_index += 1` into a `shift_next_cf` function.